### PR TITLE
Handle links with specified platform and add tests

### PR DESCRIPTION
**Pull request by timngyn**
#### Description of changes:
- Added logic to not add the `query` object if the platform was already specified in url when trying to set up an MDX link
  - Previously if we linked directly to a platform, it would add the platform as a query param (`?platform=react`)

To test:
1. Go to https://fix-specific-platform-linking.d1ywzrxfkb9wgg.amplifyapp.com/flutter/build-a-backend/existing-resources/cli/ 
2. In the callout, `Connecting a web app?...`, click on "React guide"
3. Verify that there isn't a platform query added to the url
4. Click around the site to test links
   1. Check hash link urls
      1.  https://fix-specific-platform-linking.d1ywzrxfkb9wgg.amplifyapp.com/gen2/reference/telemetry/ - The "opt out" link at the top of the page
      2.  https://fix-specific-platform-linking.d1ywzrxfkb9wgg.amplifyapp.com/react/build-a-backend/troubleshooting/upgrade-amplify-packages/#check-for-duplicate-versions` - "Check for duplicate versions" link
   2. Check "[platform]" links
      1. https://fix-specific-platform-linking.d1ywzrxfkb9wgg.amplifyapp.com/react/tools/cli/migration/transformer-migration/ - "GraphQL Transformer v2" 

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] Swift
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://docs.amplify.aws/)` 
            HTML: `<a href="https://docs.amplify.aws/">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/components/MDXComponents/MDXLink.tsx [hunk 1]
```diff
@@ -19,16 +19,25 @@ export const MDXLink = ({
     href = href.slice(0, href.indexOf('#'));
   }
 
+  // For setting up a link when passed only a hash link
   if (!href.includes('/') && href.startsWith('#')) {
     hash = href;
     href = baseURI.replace(platform, '[platform]');
   }
 
+  const decodedUri = decodeURI(href);
```
Comment (reesscot) [inline src/components/MDXComponents/MDXLink.tsx:28]: nit: keep URI uppercase for consistency
```suggestion
  const decodedURI = decodeURI(href);
```

**Review by reesscot — DISMISSED** [review]
One small nit, but LGTM

File: src/components/MDXComponents/utils/useMDXLinkHref.ts
Comment (timngyn) [inline src/components/MDXComponents/utils/useMDXLinkHref.ts:1]: Tried to refactor this out so it was easier to test, but could not get the mock for next/link to work. Should I leave it like this or merge it back into `MDXLink`?

File: src/components/MDXComponents/__tests__/MDXLink.test.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        ? `${href.pathname.replace('[platform]', href.query?.platform)}${
+            href.hash ? `${href.hash}` : ''
+          }`
+        : href;
+
+    return <a href={hrefProp}>{children}</a>;
+  };
+
+  return MockedNextLink;
+});
```
Comment (timngyn) [inline src/components/MDXComponents/__tests__/MDXLink.test.tsx:31]: I couldn't get the mock for `next/link` to work without reimplementing it in how I think it should work. This doesn't feel right and so I'm open to any feedback on this.

I tried refactoring out this the logic for setting up the `href` object so I mock `next/link` using `jest.mock('next/link', () => ({ children }) => children);` and tried to spy on it. This didn't work though, kept getting some nextjs errors

Comment (hbuchel) [reply src/components/MDXComponents/__tests__/MDXLink.test.tsx:31]: I tried a bunch of suggestions I found but couldn't get any of them to work either. I'm good with this for now.

**Review by hbuchel — APPROVED** [review]

File: jest.setup.ts [hunk 3]
```diff
@@ -1,6 +1,11 @@
+import { createContext } from 'react';
 // Setup file to extend jest-dom, referenced in packages.json under "jest"
 import '@testing-library/jest-dom';
 
+jest.doMock('next/dist/shared/lib/router-context.shared-runtime.js', () => ({
```
Comment (hbuchel) [inline jest.setup.ts:5]: Can we add a comment for why we added this?

File: src/components/MDXComponents/__tests__/MDXLink.test.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      query: {
+        platform: ''
+      },
+      asPath: ''
+    };
+  }
+};
+
+jest.mock('next/router', () => routerMock);
+//   const MockedNextLink = ({ children, href }) => {
```
Comment (hbuchel) [inline src/components/MDXComponents/__tests__/MDXLink.test.tsx:17]: Just a nit to delete this commented out stuff

**Review by hbuchel — COMMENTED** [review]
Just two small comments

File: src/components/MDXComponents/__tests__/MDXLink.test.tsx (hunk 4 above)
Comment (timngyn) [reply src/components/MDXComponents/__tests__/MDXLink.test.tsx:17]: Thank you for catching this! Sorry about that, for some reason in my vscode, the line above it was collapsed so I thought I deleted it already (it was hiding all these lines instead).

**Review by hbuchel — APPROVED** [review]

**Review by katiegoines — APPROVED** [review]

**Pull request merged**